### ISSUE TITLE
[data grid] Skeleton for loading (replace circular progress?)

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-x#1685]
## Summary 💡

The central spinner when the data load and refreshes is not very nice. 

## Motivation 🔦

Instead of a loader, we could render a skeleton. To check with the designer.

Developers can customize it based on this demo: https://material-ui.com/components/data-grid/components/#loading-overlay. 

## Benchmarks

1. https://react.carbondesignsystem.com/?path=/story/components-datatable--skeleton
<img width="832" alt="Screenshot 2021-05-17 at 18 08 25" src="https://user-images.githubusercontent.com/3165635/118521038-e7361700-b73a-11eb-98c6-9a2389ed8c78.png">

2. Toolpad. [skeleton for the data grid](https://github.com/mui/mui-toolpad/blob/82635255595a8fd7c3e0222101abd11e06b4ea0e/packages/toolpad-components/src/DataGrid.tsx#L43). It adds some randomness to the length of the bars, and it respects cell alignment. It falls back to `LinearProgress` when there already are rows. ([codesandbox](https://codesandbox.io/s/customloadingoverlaygrid-material-demo-forked-dqfb0z?file=/demo.tsx))

<img width="722" alt="Screenshot 2022-06-01 at 16 34 16" src="https://user-images.githubusercontent.com/3165635/171430050-4e66dfdb-2328-43b6-82bc-0e22439e247c.png">

One way that this concept could be taken even further is to also consider the data type of the columns and vary the randomness. e.g. for dates or numbers, the length of the content may be more uniform across rows than for text.

3. GitHub

<img width="745" alt="Screenshot 2022-06-01 at 16 37 49" src="https://user-images.githubusercontent.com/3165635/171430871-bca25f67-f8a1-4f08-9b9a-87be9fc8915c.png">

4. https://twitter.com/nots_dney/status/1540318871923486720

<img width="596" alt="SCR-20240505-sgit" src="https://github.com/mui/mui-x/assets/3165635/dee4dfa9-1e29-41a0-a3a9-7feb76284a25">

5. AG Grid https://www.ag-grid.com/react-data-grid/component-loading-cell-renderer/#skeleton-loading. Missing randomness in the length of the skeleton. Not great. See 2. on how to improve it.

**Comment by oliviertassinari** [comment]
At this point, it's an RFC, It's not clear what could yield the best UX.

**Comment by dtassone** [comment]
We should also consider when we already loaded rows and for some reason, the loading state comes up? New server call.
Should we see the skeleton?

**Comment by DanailH** [comment]
I'm looking into having such special rows for the second part of the infinite loading feature #1247  because I need to have those skeleton rows in between loaded rows. 

I was thinking that we probably need to create a special kind of row?

**Comment by DanailH** [comment]
> Just to add to the benchmark, we created a [skeleton for the data grid](https://github.com/mui/mui-toolpad/blob/82635255595a8fd7c3e0222101abd11e06b4ea0e/packages/toolpad-components/src/DataGrid.tsx#L43) in Toolpad. It adds some randomness to the length of the bars, and it respects cell alignment. ([codesandbox](https://codesandbox.io/s/customloadingoverlaygrid-material-demo-forked-dqfb0z?file=/demo.tsx))

Nice. I wonder do we need a skeleton per cell. We can use the new `colspan` feature and have it on the skeleton for the entire row.

**Comment by Janpot** [comment]
> Nice. I wonder do we need a skeleton per cell.

The way I understand it is that you'd use a skeleton instead of a spinner to minimize the visual interruption and layout shift when data is loaded. You try to resemble the expected final state as close as possible with the limited information available. That's what I aimed for with the Toolpad implementation and why I went for a skeleton per cell. I think looks better, but I wouldn't argue that it's an absolute requirement.

**Comment by oliviertassinari** [comment]
I have added another benchmark on this problem by @notsidney: https://twitter.com/nots_dney/status/1540318871923486720

**Comment by jarb99** [comment]
This would be a great feature!! Its already implemented via the Lazy loading [https://mui.com/x/react-data-grid/row-updates/#lazy-loading](url), is it possible to turn this on as a loading overlay for a certain number of rows? This would mean UI consistency in datagrid loading skeletons.

**Comment by githorse** [comment]
I would think it'd be useful to allow the user to specify custom loading screens, be that a spinner, skeleton, or something else. For instance, we have a custom branded spinner that I might want to put in the grid while it's initially loading. For something like the skeleton above, I guess, it would have to be built in. Not sure on the details of course but something like

```
<Datagrid
   renderLoading={'spinner' | 'skeleton' | ReactNode}
/>
```

There's another level of complication here: I find I usually need to distinguish between two types of "loading" states: 1) initial load, where there's no data in the grid at all to display; and 2) refreshes/reloads, where the data is there but is briefly stale. I would love to turn on a grid skeleton for the first case (initial load). For the second case, I usually implement my own "reloading" state the just dials down the opacity. (Reloading is usually fairly quick and I just want a subtle indication that something is going on, but nothing visually jarring.)

It would be cool if the DataGrid could distinguish between those two types of loading.

**Comment by coot3** [comment]
> Just to add to the benchmark, we created a [skeleton for the data grid](https://github.com/mui/mui-toolpad/blob/82635255595a8fd7c3e0222101abd11e06b4ea0e/packages/toolpad-components/src/DataGrid.tsx#L43) in Toolpad. It adds some randomness to the length of the bars, and it respects cell alignment. It falls back to `LinearProgress` when there already are rows. ([codesandbox](https://codesandbox.io/s/customloadingoverlaygrid-material-demo-forked-dqfb0z?file=/demo.tsx))

In case it helps anyone, with the MUI X v6 update, there is now a "scrollPositionChange" GridEvent.  You can make Janpots great answer more natural by using it to programatically scroll the loading skeleton overlay when the grid is overflowing and the user scrolls the datagrid while the skeleton is still showing. The skeleton cells will stay inline with the header row.  Updated example of x scrolling:

https://codesandbox.io/s/customloadingoverlaygrid-material-demo-forked-c8xt5x?file=/demo.tsx

**Comment by githorse** [comment]
[Material React Table](https://www.material-react-table.com/docs/guides/async-loading#loading-demo) does a nice job with loading IMO. (I prefer just the skeletons and not the spinner, but that's customizable.) Looks pretty much like @coot3's version.

**Comment by CezaryBD** [comment]
Bump. I am willing to use such feature, would be good to bring developers more flexibility in this area.

**Comment by vishal-kadmos** [comment]
Any update on this? Looking for similar request

**Comment by eschneor** [comment]
Any update on this?

**Comment by joserodolfofreitas** [comment]
We already have skeleton rows with lazy loading:
https://mui.com/x/react-data-grid/row-updates/#lazy-loading

Should we have a different API to use it?

**Comment by MBilalShafi** [comment]
Implemented in #13293
